matrix-bot/src/Network/Matrix/Bot/Async/Internal.hs

```diff
@@ -0,0 +1,234 @@
+module Network.Matrix.Bot.Async.Internal
```

Comment:
nit: I wonder if you had considered using [thread-supervisor](https://hackage.haskell.org/package/thread-supervisor) to implement this module? The Actor and Supervisor constructs look like a good fit for our use case, albeit the use a concret IO.

Reply:
I had not. But it looks suitable for the job. Unfortunately, some of its functionality is not usable for us, because it works on `IO`, but we need it to work on arbitrary `MonadIO`. I think this could be fixed by a patch to thread-supervisor, but I think its best if we user it as-is for now and maybe consolidate the functionality in both packages later.

Reply:
Done, as described above.